lib/task_005_lib.rb

```diff
@@ -0,0 +1,95 @@
+module Task005
+  module_function
+
+  class Point
+    attr_reader :x, :y
+
+    def self.from_input(point)
+      x, y = point.split(',').map(&:to_i)
+      Point.new(x, y)
+    end
+
+    def initialize(x, y)
+      @x = x
+      @y = y
+    end
+  end
+
+  class Line
+    attr_reader :point1, :point2
+
+    def self.from_input(line)
+      point1, point2 = line.split(' -> ').map { |point| Point.from_input(point) }
+      Line.new(point1, point2)
+    end
+
+    def initialize(point1, point2)
+      @point1 = point1
+      @point2 = point2
+    end
+
+    def horizontal_or_vertical?(line)
+      line.point1.x == line.point2.x || line.point1.y == line.point2.y
+    end
+
+    def degree_45?(line)
+      (line.point1.x - line.point2.x).abs == (line.point1.y - line.point2.y).abs
+    end
+
+    def points_on_line(line)
+      if horizontal_or_vertical?(line)
+        points = []
+        x1, x2 = [line.point1.x, line.point2.x].sort
+        y1, y2 = [line.point1.y, line.point2.y].sort
+        (x1..x2).each do |x|
+          (y1..y2).each do |y|
+            points << [x, y]
+          end
+        end
+      elsif degree_45?(line)
+        points = []
+        x1, x2 = [line.point1.x, line.point2.x].sort
+        y1, y2 = [line.point1.y, line.point2.y].sort
+
+        x = line.point1.x - line.point2.x < 0 ? (x1..x2).each.to_a : (x1..x2).each.to_a.reverse
+        y = line.point1.y - line.point2.y < 0 ? (y1..y2).each.to_a : (y1..y2).each.to_a.reverse
+
+        x.size.times do |i|
+          points << [x[i], y[i]]
+        end
+      end
+      points
+    end
+  end
+
+  class Diagram
+    attr_accessor :diagram
+
+    def initialize(diagram = {})
+      @diagram = diagram
+    end
+
+    def mark_line(line)
+      line.points_on_line(line).each do |x, y|
+        @diagram.key?([x, y]) ? @diagram[[x, y]] += 1 : @diagram.store([x, y], 1)
+      end
+    end
+
+    def mark_horizontal_and_vertical(lines)
+      lines.each { |line| line.horizontal_or_vertical?(line) ? mark_line(line) : nil }
+      @diagram
+    end
+
+    def mark_horizontal_vertical_45_degree(lines)
+      lines.each do |line|
+        line.horizontal_or_vertical?(line) ? mark_line(line) : nil
+        line.degree_45?(line) ? mark_line(line) : nil
+      end
+      @diagram
+    end
+  end
```

Comment:
I am not sure if it should be in the `Diagram` class. I would probably leave it in some external function that calculates this task. But that's just a feeling, not a strong opinion.

Reply:
I know what do you mean, I was thinking from begin about this what is going with Diagram Class. In this case I'm only use Class Diagram to create `@diagram`, I mean there is no work on object/instance `diagram`, only creation of `@diagram`.
I'm not sure if this is clear what I mean, but this two methods are doing that job.

Reply:
If you look at the code, you can see that instances of `Point` , `Lines` , are been used but instance of `Diagram` nowhere. 
That was my problem with Diagram Class from begin, but if I move this two methods out of `Diagram` then it  will be used, but I think it is still something wrong.

Reply:
I mean it will be empty {} before use this methods. Also to get `@diagram` need to call `diagram.diagram`.

Reply:
I think that I already have class that I need, it is class `Hash,` and just use methods for `Hash.` Other way is to rebuild class `Diagram,` because in actual state class `Diagram` looks redundant.

Reply:
Great comments! So the Hash would obviously do the work here. But I would also try the Diagram class first because I believe there is a better solution :)

> I mean it will be empty {} before use this methods. Also to get @diagram need to call diagram.diagram.

This is a great observation and you started going in the right direction. Or at least the direction that is good in my opinion, and at the end of this refactor you will be able to judge it yourself.

I agree that this API `diagram.diagram` does not look good ;)

The question I would ask is: Do you need to expose `@diagram` to the outside? If you think "yes" - then why?

In the other words: can you remove `attr_reader :diagram` and stop returning `@diagram` from the methods and still solve the task?

Reply:
Yes, I can do that :)  and it works fine :), but I just left  `attr_reader :diagram` , need it for spec file, `contain_exactly` have some problem to work with